### PR TITLE
Make EC2 AWS API endpoint configurable in operator

### LINKED ISSUE #12620: Make AWS API endpoint used by cilium-operator configurable
## Proposal / RFE

cilium-operator need to talk to the AWS API endpoints for things like ENI allocation. 

In some scenarios, we may need to talk to custom API endpoints as described in the `Using Custom Endpoints` section here: https://docs.aws.amazon.com/sdk-for-go/v2/api/aws/endpoints/

For example, we might want to use the FIPS endpoints listed here: https://aws.amazon.com/compliance/fips/

### PR DESCRIPTION
**Pull request by tklauser**
Add a new --ec2-api-endpoint operator option which allows to specify
a custom AWS API endpoints for the EC2 service. One possible use-case
for this is the usage of FIPS endpoints, see
https://aws.amazon.com/compliance/fips/. For example, to use API
endpoint ec2-fips.us-west-1.amazonaws.com, the AWS operator can be
called using:
    
    cilium-operator-aws --ec2-api-endpoint=ec2-fips.us-west-1.amazonaws.com
    
Note: for now this only allows to configure the API endpoint for the EC2
service. It would be possible to make this more generic and configure the
endpoint per service, e.g. using a map of strings option

    cilium-operator-aws --aws-api-endpoints=ec2=ec2-fips.us-west-1.amazonaws.com,otherservice=other-ep

But as far as I can see we're currently only using the EC2 API endpoint in
the operator.

The first commit is preparatory go vendoring of the latest AWS Go SDK.
The second commit fixes a small typo I noticed while working on this.
The third commit contains the actual implementation.

Fixes #12620

### CODE REVIEW

**Comment by tklauser** [comment]
test-me-please

**Comment by tklauser** [comment]
retest-net-next

**Comment by tklauser** [comment]
test-me-please

**Review by christarazi — APPROVED** [review]

**Review by aanm — APPROVED** [review]

File: pkg/aws/endpoints/resolver.go
Comment (fristonio) [inline pkg/aws/endpoints/resolver.go:37]: Do we need to prepend the `https://` prefix here? I see that in the docs you linked we can provide URL without any scheme, like:

```
aws.Endpoint{
        URL:           "s3.custom.endpoint.com",
        SigningRegion: "custom-signing-region",
}
```

I think it might be good to use the endpoint provided by the user as it is without adding a scheme as we don't do any other validation here. So let's say if a user provides a URL with `https://` already prefixed this will not work. If we don't add a scheme user can provide URL either without the scheme or with the scheme.

**Review by fristonio — APPROVED** [review]
Looks great :partying_face: 
A minor suggestion which does not require a second review.

File: pkg/aws/endpoints/resolver.go
Comment (tklauser) [reply pkg/aws/endpoints/resolver.go:37]: The linked docs are a bit misleading in that respect. In my tests, the endpoint URL was only working if a scheme was provided. Given that `https://` is the only one that worked here, I thought it's easier to just hard-code this and let the user only specify the hostname.

Comment (fristonio) [reply pkg/aws/endpoints/resolver.go:37]: In that case, it makes sense, for now, to go with this. This shouldn't be a problem as in the example of the command-line flag we do provide the right kind of endpoint we are expecting. We can always circle back to this later if anyone faces the issue.

Comment (tklauser) [reply pkg/aws/endpoints/resolver.go:37]: Sounds good, thank you for the review @fristonio!

**Pull request merged**
